github.go

```diff
@@ -0,0 +1,267 @@
+	)
+	if err != nil {
+		if resp.StatusCode == 404 {
+			return nil, "", errFileNotFound
+		}
+		return nil, "", err
+	}
+
+	decoded, err := contents.GetContent()
+	return []byte(decoded), *contents.SHA, err
+}
+
+func (gh GitHub) writeFile(filename string, sha string, contents []byte) error {
+	if err := gh.ensureClient(); err != nil {
+		return err
+	}
+
+	var err error
+	var writeFunc func(context.Context, string, string, string, *github.RepositoryContentFileOptions) (*github.RepositoryContentResponse, *github.Response, error)
+	opts := &github.RepositoryContentFileOptions{
+		Message: github.String(fmt.Sprintf("[checkup] store %s [ci skip]", filename)),
+		Content: contents,
+		Committer: &github.CommitAuthor{
+			Name:  &gh.CommitterName,
+			Email: &gh.CommitterEmail,
+		},
+	}
+
+	if gh.Branch != "" {
+		opts.Branch = &gh.Branch
+	}
+
+	// If no SHA specified, then create the file.
+	// Otherwise, update the file at the specified SHA.
+	if sha == "" {
+		writeFunc = gh.client.Repositories.CreateFile
+		fmt.Printf("creating %s on branch=%s\n", gh.fullPathName(filename), gh.Branch)
+	} else {
+		opts.SHA = github.String(sha)
+		writeFunc = gh.client.Repositories.UpdateFile
+		fmt.Printf("updating %s on branch=%s\n", gh.fullPathName(filename), gh.Branch)
+	}
+
+	_, _, err = writeFunc(
+		context.Background(),
+		gh.RepositoryOwner,
+		gh.RepositoryName,
+		gh.fullPathName(filename),
+		opts,
+	)
+	return err
+}
+
+// deleteFile deletes a file from a Git tree and returns the new SHA for the ref
+// and any applicable errors. If an error occurs, the input SHA is returned along
+// with the error.
+func (gh GitHub) deleteFile(filename string, sha string) (string, error) {
+	if err := gh.ensureClient(); err != nil {
+		return "", err
+	}
+
+	if sha == "" {
+		return "", errFileNotFound
+	}
+
+	commit, _, err := gh.client.Repositories.DeleteFile(
+		context.Background(),
+		gh.RepositoryOwner,
+		gh.RepositoryName,
+		gh.fullPathName(filename),
+		&github.RepositoryContentFileOptions{
+			Message: github.String(fmt.Sprintf("[checkup] delete %s [ci skip]", filepath.Base(filename))),
+			SHA:     github.String(sha),
+			Committer: &github.CommitAuthor{
+				Name:  &gh.CommitterName,
+				Email: &gh.CommitterEmail,
+			},
+		},
+	)
+	if err != nil {
+		return sha, err
+	}
+
+	return *commit.Commit.SHA, nil
+}
+
+func (gh GitHub) readIndex() (map[string]int64, string, error) {
+	index := map[string]int64{}
+
+	contents, sha, err := gh.readFile(indexName)
+	if err != nil && err != errFileNotFound {
+		return nil, "", err
+	}
+	if err == errFileNotFound {
+		return index, "", nil
+	}
+
+	err = json.Unmarshal(contents, &index)
+	return index, sha, err
+}
+
+func (gh GitHub) writeIndex(index map[string]int64, sha string) error {
+	contents, err := json.Marshal(index)
+	if err != nil {
+		return err
+	}
+
+	return gh.writeFile(indexName, sha, contents)
+}
+
+// Store stores results on filesystem according to the configuration in fs.
```

Comment:
Make sure the godoc and other comments reflect this code, not the code it was borrowed from ;)